mu-plugins/enforce-2fa.php

```diff
@@ -0,0 +1,438 @@
+		}
+		add_action( 'set_current_user', 'e2fa_enforce_two_factor_plugin' );
+	}
+}
+
+/**
+ * Filter Caps
+ *
+ * Remove caps for users without two-factor enabled so they are treated as a Contributor.
+ *
+ * @param array  $caps    The user's actual capabilities.
+ * @param string $cap     Capability name.
+ * @param int    $user_id The user ID.
+ * @param array  $args    Adds the context to the cap. Typically the object ID.
+ *
+ * @return array
+ */
+function e2fa_two_factor_filter_caps( $caps, $cap, $user_id, $args ) {
+
+	if ( e2fa_is_two_factor_forced() ) {
+		// Use a hard-coded list of caps that give just enough access to set up 2FA.
+		$subscriber_caps = array(
+			'read',
+			'level_0',
+		);
+
+		// You can edit your own user account (required to set up 2FA).
+		if ( 'edit_user' === $cap && ! empty( $args ) && (int) $user_id === (int) $args[0] ) {
+			$subscriber_caps[] = 'edit_user';
+		}
+
+		// WooCommerce caps to check.
+		$woocommerce_caps = array(
+			'edit_posts',
+			'manage_woocommerce',
+			'view_admin_dashboard',
+		);
+
+		// Track whether or not we've already granted this user wp-admin access based on WC standards.
+		static $user_should_have_wc_admin_access = false;
+
+		// If we haven't granted access yet, and this $cap is a WC cap to check.
+		if ( ! $user_should_have_wc_admin_access && in_array( $cap, $woocommerce_caps, true ) ) {
+
+			// If this user has this $cap and it's `true`, grant this user wp-admin access.
+			if ( isset( wp_get_current_user()->allcaps[ $cap ] ) && true === wp_get_current_user()->allcaps[ $cap ] ) {
+				$user_should_have_wc_admin_access = true;
+				add_filter( 'woocommerce_prevent_admin_access', '__return_false' );
+			}
+		}
+		if ( ! in_array( $cap, $subscriber_caps, true ) ) {
+			return array( 'do_not_allow' );
+		}
+	}
+	return $caps;
+}
+
+/**
+ * Ristrict block editor screen.
+ *
+ * @return bool
+ */
+function e2fa_should_show_notice_on_current_screen() {
+
+	$screen = get_current_screen();
+
+	// Don't show on the "Edit Post" screen as it interferes with the Block Editor.
+	if ( $screen->is_block_editor() ) {
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Admin Notice for 2FA.
+ *
+ * @return void
+ */
+function e2fa_two_factor_admin_notice() {
+
+	// If 2FA is not forced, return early.
+	if ( ! e2fa_is_two_factor_forced() ) {
+		return;
+	}
+
+	if ( ! e2fa_should_show_notice_on_current_screen() ) {
+		return;
+	}
+
+	?>
+	<div id="e2fa-error" class="notice-error wrap clearfix" style="align-items: center;background: #ffffff;border-left-width:4px;border-left-style:solid;border-radius: 6px;display: flex;margin-top: 30px;padding: 30px;line-height: 2em;">
+			<div class="dashicons dashicons-warning" style="display:flex;float:left;margin-right:2rem;font-size:38px;align-items:center;margin-left:-20px;color:#ffb900;"></div>
+			<div>
+				<p style="font-weight:bold; font-size:16px;">Two Factor Authentication is required to edit content on this site.</p>
```

Comment:
Do we need to escape and translate the strings here?

Reply:
Added esc_html_e without text domain on strings.